### PR TITLE
Add support for 4K response data

### PR DESCRIPTION
**Pull request by xyan264**
New function `switchtec_nvme_admin_passthru` from switchtec-user library supports 4K response data from SRIOV devices

### CODE REVIEW

File: plugins/microchip/switchtec-nvme-device.c [hunk 1]
```diff
@@ -48,8 +49,20 @@ struct switchtec_device_manage_nvme_rsp
...
+};
 
-int pax_nvme_submit_admin_passthru(int fd, struct nvme_passthru_cmd *cmd)
+struct switchtec_adm_passthru_nvme_rsp
+{
+	uint32_t nvme_cqe[4];
+	uint8_t nvme_data[4096];
+};
+
+int pax_nvme_submit_admin_passthru_v1(int fd, struct nvme_passthru_cmd *cmd)
```
Comment (kelvin-cao) [inline plugins/microchip/switchtec-nvme-device.c:65]: _v1 seems not clear, how about _1k_rsp?

File: plugins/microchip/switchtec-nvme-device.c [hunk 2]
```diff
@@ -100,6 +113,61 @@ int pax_nvme_submit_admin_passthru(int fd, struct nvme_passthru_cmd *cmd)
...
+
+	memcpy(req.nvme_data, (void *)cmd->addr, data_len);
+	data_len = data_len + 16 * 4;
+
+	rsp_len = (sizeof(rsp.nvme_cqe) + cmd->data_len);
+
+	ret = switchtec_nvme_admin_passthru(pax->dev, pax->pdfid,
+					    data_len, &req,
+					    &rsp_len, &rsp);
+	if (ret && errno == EBADMSG) {
```
Comment (kelvin-cao) [inline plugins/microchip/switchtec-nvme-device.c:149]: EBADMSG from firmware? Should be UNSUPPORTED?

File: plugins/microchip/switchtec-nvme-device.c [hunk 3]
```diff
@@ -100,6 +113,61 @@ int pax_nvme_submit_admin_passthru(int fd, struct nvme_passthru_cmd *cmd)
 	return status;
 }
 
+int pax_nvme_submit_admin_passthru(int fd, struct nvme_passthru_cmd *cmd)
```
Comment (kelvin-cao) [inline plugins/microchip/switchtec-nvme-device.c:116]: We should have another _4k_rsp function for the new implementation, and this function just try _4k_rsp and 1k_rsp and return.

File: plugins/microchip/switchtec-nvme-device.c [hunk 4]
```diff
@@ -48,8 +49,20 @@ struct switchtec_device_manage_nvme_rsp
 	uint8_t nvme_data[SWITCHTEC_DEVICE_MANAGE_MAX_RESP - (4 * 4)];
 };
 
+struct switchtec_adm_passthru_nvme_req
+{
```
Comment (kelvin-cao) [inline plugins/microchip/switchtec-nvme-device.c:53]: '{' should be in the same line of struct

File: plugins/microchip/switchtec-nvme-device.c [hunk 5]
```diff
@@ -48,8 +49,20 @@ struct switchtec_device_manage_nvme_rsp
...
+	uint32_t nvme_sqe[16];
+	uint8_t nvme_data[SWITCHTEC_NVME_ADMIN_PASSTHRU_MAX_DATA_LEN -
+				(16 * 4)];
+};
 
-int pax_nvme_submit_admin_passthru(int fd, struct nvme_passthru_cmd *cmd)
+struct switchtec_adm_passthru_nvme_rsp
+{
```
Comment (kelvin-cao) [inline plugins/microchip/switchtec-nvme-device.c:60]: ditto

File: plugins/microchip/switchtec-nvme-device.c [hunk 6]
```diff
@@ -100,6 +113,61 @@ int pax_nvme_submit_admin_passthru(int fd, struct nvme_passthru_cmd *cmd)
...
+	data_len = data_len + 16 * 4;
+
+	rsp_len = (sizeof(rsp.nvme_cqe) + cmd->data_len);
+
+	ret = switchtec_nvme_admin_passthru(pax->dev, pax->pdfid,
+					    data_len, &req,
+					    &rsp_len, &rsp);
+	if (ret && errno == EBADMSG) {
+		/* if device does not support new function, */
+		/* call legacy function instead */
```
Comment (kelvin-cao) [inline plugins/microchip/switchtec-nvme-device.c:151]: Normally we do multi-line comment like this 
```
/*
 * This is an example of multi-line comment. we'll wrap when we reach
 * 80 characters.
 */
```

File: plugins/microchip/switchtec-nvme-device.c (hunk 2 above)
Comment (xyan264) [reply plugins/microchip/switchtec-nvme-device.c]: Yes this is a problem. Let's discuss.

File: plugins/microchip/switchtec-nvme-device.c [hunk 7]
```diff
@@ -100,6 +109,70 @@ int pax_nvme_submit_admin_passthru(int fd, struct nvme_passthru_cmd *cmd)
...
+int pax_nvme_submit_admin_passthru(int fd, struct nvme_passthru_cmd *cmd)
+{
+	int ret;
+
+	ret = pax_nvme_submit_admin_passthru_4k_rsp(fd, cmd);
+	if (ret)
+		ret = pax_nvme_submit_admin_passthru_1k_rsp(fd, cmd);
+
+	if (ret)
+		switchtec_perror("pax_nvme_submit_admin_passthru");
```
Comment (kelvin-cao) [inline plugins/microchip/switchtec-nvme-device.c:171]: We should not do the error message output in the API, instead we leave it to the callers.

**Comment by xyan264** [comment]
@kelvin-cao Code updated with the following changes:
- removed error prints from `pax_nvme_submit_admin_passthru`
- added `errno` checking:
```	ret = pax_nvme_submit_admin_passthru_4k_rsp(fd, cmd);
	if (ret && errno == INVALID_MRPC_CMD)
		ret = pax_nvme_submit_admin_passthru_1k_rsp(fd, cmd);
```

**Review by kelvin-cao — APPROVED** [review]

**Pull request merged**